### PR TITLE
feat: Add variants metrics to SDK

### PR DESCRIPTION
**Pull request by thomasheartman**
This PR adds support for variant metrics to the Rust SDK. It's about 95 lines of implementation and 150 lines of tests.

Variant metrics should be counted whenever `get_variant` is called:
- if a feature doesn't have variants, register the `disabled` variant
- if a feature is disabled, register the `disabled` variant
- if a feature doesn't exist, it should still register the `disabled` variant.
- otherwise, register the variant that was returned

## Discussion points

I've implemented the variant metrics on each feature instance as a `DashMap` (and thus also added a new crate).
Currently, each feature will get a new DashMap initialized. It gets initialized with a capacity of 0.
It _may_ be worth making the DashMap an Option and only initializing it when it's requested, but that might require a fair 
bit of trickery (and suggestions on how to do it would be very welcome).

### Tests 

I've added tests to ensure that metrics are being counted and that the conversion from CachedFeature to ToggleMetrics works as expected. I've added new tests for these, but the tests for variant counting **could** be added to the existing get_variant tests. However, I felt that would muddy the tests, so I preferred creating new tests.

I have not tested the serde serialization output, because I don't think that would add much value. But I'd be happy to put in a test for that too if you think it would be useful.

### Benches

Of course, benchmarks are a fickle thing, but I figured they'd be useful to make sure there's no giant blunders. Running it locally on my machine against main, I found that most benchmarks have no change or it is within the noise threshold. Three benches indicated performance improved, and three indicated that it had regressed.

Most of the improvements and regressions are =< ~6%, so I don't know how much stock to put in them. However, one regression, `batch/parallel unknown-features(str)` reported a throughput decrease of 24%, so that may be significant.

#### Performance improvements

While I don't see how this would improve performance in any way, the following benches were reported as improved:

```
single_call/single_call(str)
                        time:   [253.58 ns 254.46 ns 255.51 ns]
                        thrpt:  [3.9137 Melem/s 3.9299 Melem/s 3.9435 Melem/s]
                 change:
                        time:   [-3.5225% -2.9598% -2.3842%] (p = 0.00 < 0.05)
                        thrpt:  [+2.4425% +3.0501% +3.6511%]
                        Performance has improved.

batch/parallel distinct-features(enum)
                        time:   [24.728 ms 24.762 ms 24.838 ms]
                        thrpt:  [16.104 Melem/s 16.154 Melem/s 16.176 Melem/s]
                 change:
                        time:   [-6.5260% -5.8273% -5.1205%] (p = 0.00 < 0.05)
                        thrpt:  [+5.3969% +6.1878% +6.9816%]
                        Performance has improved.

batch/parallel unknown-features(enum)
                        time:   [2.9772 ms 2.9899 ms 3.0029 ms]
                        thrpt:  [133.20 Melem/s 133.79 Melem/s 134.35 Melem/s]
                 change:
                        time:   [-6.5641% -6.0584% -5.5674%] (p = 0.00 < 0.05)
                        thrpt:  [+5.8956% +6.4491% +7.0252%]
                        Performance has improved.
```

#### Performance regressions

The following regressions were reported:

```
batch/single thread(str)
                        time:   [12.815 ms 12.892 ms 12.974 ms]
                        thrpt:  [3.8537 Melem/s 3.8783 Melem/s 3.9016 Melem/s]
                 change:
                        time:   [+1.3055% +2.0763% +2.8471%] (p = 0.00 < 0.05)
                        thrpt:  [-2.7683% -2.0341% -1.2887%]
                        Performance has regressed.

batch/parallel same-feature(str)
                        time:   [28.019 ms 28.085 ms 28.170 ms]
                        thrpt:  [14.199 Melem/s 14.242 Melem/s 14.276 Melem/s]
                 change:
                        time:   [+1.2655% +1.8355% +2.4361%] (p = 0.00 < 0.05)
                        thrpt:  [-2.3782% -1.8025% -1.2497%]
                        Performance has regressed.

batch/parallel unknown-features(str)
                        time:   [4.2516 ms 4.2622 ms 4.2794 ms]
                        thrpt:  [93.470 Melem/s 93.849 Melem/s 94.082 Melem/s]
                 change:
                        time:   [+30.189% +31.927% +33.541%] (p = 0.00 < 0.05)
                        thrpt:  [-25.116% -24.200% -23.188%]
                        Performance has regressed.
```

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -142,10 +143,48 @@ pub struct CachedFeature {
...
+            .and_modify(|count| {
+                count.fetch_add(1, Ordering::Relaxed);
+            })
+            .or_insert(AtomicU64::new(1));
+    }
+}
+
+impl From<&CachedFeature> for ToggleMetrics {
+    fn from(feature: &CachedFeature) -> Self {
+        fn convert_variants_count(original: &DashMap<String, AtomicU64>) -> HashMap<String, u64> {
```
Comment (sighphyre) [inline src/client.rs:164]: I want to say this could be more idiomatically expressed as:

```fn convert_variants_count(original: &DashMap<String, AtomicU64>) -> HashMap<String, u64> {
    original
        .iter()
        .map(|item| {
            let (key, value) = item.pair();
            let count = value.load(Ordering::Relaxed);
            (key.clone(), count)
        })
        .collect()
}```

Not tested, just eyeballed. And this may be a matter of taste so up to you

File: src/client.rs [hunk 2]
```diff
@@ -348,10 +387,29 @@ where
                             features: EnumMap::default(),
                             str_features: HashMap::new(),
                         };
+
+                        fn clone_variants_map(
```
Comment (sighphyre) [inline src/client.rs:391]: Same as above

File: src/client.rs (hunk 1 above)
Comment (thomasheartman) [reply src/client.rs:164]: Nah, I like that a lot! Super catch 🙌🏼

**Review by sighphyre — APPROVED** [review]
I like it! LGTM, probably worth checking what the benchmarks say before and after this PR but I don't see anything that _should_ cause major performance regressions

File: src/client.rs [hunk 3]
```diff
@@ -461,7 +480,9 @@ where
             Some(cache) => cache,
         };
         let enabled = cache.is_enabled_str(feature_name, Some(context), false, &self.cached_state);
+        let feature = &cache.str_features.get(feature_name);
```
[the remark this answers is not in the document]
Comment (thomasheartman) [reply src/client.rs:521]: As mentioned in the PR description, most of it is more or less the same. The one instance that does see significant changes is parallel unknown-features(str). When you try and get a feature that doesn't exist, we initialize a new dashmap for the feature now. I suspect that this line is what causes that.

**Comment by thomasheartman** [comment]
@sighphyre As mentioned in the comment, I  do think that the one significant regression may be related to dashmap initialization. I'll see if I can figure a way around it and if it changes anything.

As for clippy: yeah, I got those lints locally as well, but I didn't want to touch them because those lines were unrelated to the PR. Happy to fix them up, though!

File: src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sighphyre) [reply src/client.rs:521]: I'm a banana, thanks, I didn't read the updated description. 24% is significant but that should be cold path so I'm not that worried. Everything else looks like statistical noise to me

**Comment by thomasheartman** [comment]
Edit: I was slightly off earlier. When you check `is_enabled_str` for a feature that doesn't exist, **that's** when we initialize a dashmap. That's gonna be harder to get around.

File: src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thomasheartman) [reply src/client.rs:521]: > 24% is significant but that should be cold path

Yeah, this is my thinking as well. I don't imagine you'd check _new_ unknown features in a loop.

> Everything else looks like statistical noise to me

Agreed. So nothing in most cases, but a marked regression for what should be a very cold path.

I did some digging and realized that the initialization of the dashmap happens when you check `is_enabled_str` with an unknown feature. That happens regardless of whether this line was moved earlier or not. 

I've got an idea that might work, however: if we store unknown variant stats on the cache itself, then we'll save some initialization work. That might work. The tradeoff is that you'll always carry that thing around. I'll see what I can do.

[the remark this answers is not in the document]
Comment (thomasheartman) [reply src/client.rs:521]: Thinking about this a bit more, I don't think that idea will work either.

The core of this is around [this clause in `is_enabled_str`](https://github.com/Unleash/unleash-client-rust/blob/2419b778eba71ea038958f9267b2c1de6da27c1b/src/client.rs#L382-L436).

What happens is that if the feature is unknown, we currently:
- create a new cached state from the old state, cloning old features and adding them
- creating a new feature for the unknown feature and inserting that into the cache
- replacing the old cached state with the new state

My thought was that if the unknown str feature doesn't need to initialize its own map, then we could save some there. However, if unknown str features could **not** have variant counts, then they have to be different from other features. This would likely cause a larger restructuring.

Maybe (_maybe_!) we could avoid storing unknown str features as actual features and just store them in a map (I think they're only created for metrics' sake), but that would also be a lot of restructuring work and I don't think it's worth it (at least not for now).

In short, I think the regressions for the cold path here are acceptable and vote we go forward with it.

Let me know if you disagree.

**Comment by rbtcollins** [comment]
Hey, so a few thoughts.

Firstly, great to see work coming in! Forgive my commenting just on the description, its all the time I have today.

I presume that this is not running conformance tests - we kind of need to fix up https://github.com/Unleash/unleash-client-rust/pull/42 and merge it and then repeat for the other commits to the conformance test suite.

Being up to date with that suite is really important to me because it lets us see that the official API works with new features, and I'd really encourage work on that before new features - because we have to manually test all new features and can't lean on the conformance test suite as much as we might otherwise.

Secondly, on dashmap - I need to look at the code I think. I think a rwlock in every get_variant call is a bit much - the basic ideal I think we're aiming for is 'smart if's', which means (to me) that it shouldn't matter how often unleash is called into, its not meaningfully worse than using 'if some_manual_feature_constant'. In particular the cold path concerns me - if the API is down, or someone deletes the toggle from the API, we shouldn't suddenly slow things down. Admittedly this SDK has a lot of headroom, but thats only because we've paid attention to overheads all the way :>.

In terms of code alternatives, I think generating an efficient lock-free structure from the API document, putting in the RCU ARC, and then indexing into it should be very efficient, lock tree, and quite easy to code.

**Comment by thomasheartman** [comment]
Closing this in favor of #60
